src/kani-verifier/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This crate includes two "proxy binaries": `kani` and `cargo-kani`.
+//! These are conveniences to make it easy to:
+//!
+//! ```bash
+//! cargo install --locked kani-verifer
+//! ```
+//!
+//! Upon first run, or upon running `cargo-kani setup`, these proxy
+//! binaries will download the appropriate Kani release bundle and invoke
+//! the "real" `kani` and `cargo-kani` binaries.
+
+#![warn(clippy::all, clippy::cargo)]
+
+use std::env;
+use std::ffi::OsString;
+use std::os::unix::prelude::CommandExt;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use anyhow::{bail, Context, Result};
+
+/// Comes from our Cargo.toml manifest file. Must correspond to our release verion.
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+/// Set by our `build.rs`, reflects the Rust target triple we're building for
+const TARGET: &str = env!("TARGET");
+
+/// Typically `~/.kani/kani-1.x/`
+fn kani_dir() -> PathBuf {
+    home::home_dir()
+        .expect("Couldn't find home dir?")
+        .join(".kani")
+        .join(format!("kani-{}", VERSION))
+}
+
+/// The filename of the release bundle
+fn download_filename() -> String {
+    format!("kani-{}-{}.tar.gz", VERSION, TARGET)
+}
+
+/// Helper to find the download URL for this version of Kani
+fn download_url() -> String {
+    let tag: &str = &format!("kani-{}", VERSION);
+    let file: &str = &download_filename();
+    format!("https://github.com/model-checking/kani/releases/download/{}/{}", tag, file)
+}
+
+/// Effectively the entry point (i.e. `main` function) for both our proxy binaries.
+pub fn proxy(bin: &str) -> Result<()> {
+    // In an effort to keep our dependencies minimal, we do the bare minimum argument parsing
+    let args: Vec<_> = std::env::args_os().collect();
+    if args.len() >= 2 && args[1] == "setup" {
+        if args.len() >= 4 && args[2] == "--use-local-bundle" {
+            setup(Some(args[3].clone()))
+        } else {
+            setup(None)
+        }
+    } else {
+        if !appears_setup() {
+            setup(None)?;
+        }
+        exec(bin)
+    }
+}
+
+/// Fast check to see if we look setup already
+fn appears_setup() -> bool {
+    kani_dir().exists()
```

Comment:
I do have a question about versions. What will happen if the user has `kani` installed but they are trying to use a different version. I think we want a way to check the version and allow user to upgrade or even downgrade their current setup.

I'm ok for now to only support one version, but it is also something that we might want to deal with eventually.

Reply:
The version is currently baked in to the proxy. This one for example is looking for 0.1 (set as the package version in Cargo.toml).

So it's finding `~/.kani/kani-0.1/` and everything is under that.

Reply:
Ah, I totally missed that. Thanks!